UI: Scene item overflow, handle/outline change and hover outline

VodBox opened the pull request:
![Scene Item Overflow](https://i.imgur.com/ofxCG14.png)

These changes should hopefully make the preview a little easier to read, as well as making the handles more familiar to people who use image/video editors.

SuslikV commented:
From the screenshot above is not clear what object is selected right now.
And "Highlight the out of bounds objects" is not an optional feature.

MisterTeeXD commented:
I would love to see this in the next update as this could solve quite a problem, not to mention those red boxes looks very nice, Well done :D

dodgepong commented:
@SuslikV I think the screenshot is a bit misleading -- `Color Source`, the first `Color Source 2`, and `Color Source 3` are all selected (with `Color Source` being the most-recently selected). The fact that the top `Color Source` has a color applied to the list item makes it hard to observe that. You'll notice that the second `Color Source 2` is not selected, and as such does not have the resize handles around it.

That screenshot does seem to make the handles look a little on the big side to me. If they have the same width as the diameter of the old circular handles, then maybe they should be shrunk a little due to the fact that they have extra corners now, making them appear a bit bigger. Could just be me, though.

VodBox commented:
If a scene item is selected and visible, an overflow will be rendered. If either is not true, it will not render an overflow. That's what my screenshot was trying to show, as the two on the top left are selected and visible, the bottom is left is selected but not visible, and the right one is not selected.

To be honest, I don't see why this would need to be made an optional feature, if it's only going to happen for selected items anyway?

As far as handle size, it would seem I left the handle size bigger from when I was messing with different circle sizes. I will pick a more reasonable size and force push that sometime today.

Fenrirthviti commented:
I also agree that the handles are a bit too big. A few px reduction in size might be a good idea. Looks great otherwise!

SuslikV commented:
What the sense in this feature then? If you need to select EACH object that may already lying out of preview window (whole object may be placed out of preview to be animated later with the plugins). I don't get it. Is this some kind of "eye candy"? Circles to squares? My oh my, I thought that this is some kind of helper (there is no dummy objects, gird, rulers etc. in obs).

@dodgepong I missed the commit meaning, and in my OBS Studio if source selected in the Dark theme it highlighted in Sources list too... Is this new bug with the selection?

@VodBox is this Dark theme screenshot?

VodBox commented:
Given that without glasses, A: I can't see the border of an item outside the canvas to know where a source is, and B: can't see the circle handles, but can tell where the square handles are, I'd say there's plenty of sense, and is hardly just eye candy.

The reason to only have the overflow showing with selected items, is that the whole point is to make editing more readable. The selection still works if you select the item in the sources list, so you don't have to be clicking outside the canvas at nothing to find it, so I don't know why you're making selecting the items seem like it's such a hurdle.

Additionally, it speeds up recognition of both where a source is, and how big it is. For example, if you've scaled up a scene item, and the source ups in resolution (such as with webcams), then the overflow is immediately alerting you to the fact that the size of the scene item extends well beyond the canvas, and in a much more clear way than the 1px thick red border does.

Yes, this is a dark theme screenshot.

SuslikV commented:
Here is how items is highlighted in the _Sources_ list for the Dark theme (all but "Color Source 5" is selected in the Preview):

![obs studio v22 0 1 ui screnshot selection](https://user-images.githubusercontent.com/19683044/51083657-69a30c00-1726-11e9-9c4c-34bd0a50b044.png)

By the way, the feature itself was discussed long time ago.

Warchamp7 commented:
@SuslikV What version of Windows are you on? That highlight is not how Dark theme looks on Windows 10, and I wonder if there's a qss bug or something

![image](https://user-images.githubusercontent.com/1554753/51141611-bf0b1600-1817-11e9-83b4-c8a737214806.png)

This is 22.0.2 with Color Source 2 selected. The highlight is a slightly darker color than the background. It's not very noticeable, and perhaps that should be adjusted. It's not related to this PR though

Dillons changes are entirely within the preview area

VodBox commented:
@DDRBoxman Before committing my changes, some others and I went through a bunch of programs to look at their handle behavior. Of the ones that were square handles, none of them rotated.

VodBox commented:
I added options to Settings around how the overflow should be shown, and fixed handle sizes being too large. Here's a (hopefully) clearer screenshot.

![Screenshot](https://i.imgur.com/B3R7U3A.png)

SuslikV commented:
@Warchamp7 , screenshot from the Win7, and it seems that this is normal behavior, because selection color for the this complex widget with custom painting(!) never was customized in the Dark theme, so it may be different for all systems. I can say that for some colors it is hard to do without the code change (because the text color of the label in non-/selected states is always the same, and the item's own text element not used).

Text and objects of interest for customization in non-/selected states (for all obs versions):
```
SourceTree QLabel,
SourceTree QCheckBox,
SourceListWidget QCheckBox,
SourceListWidget QLabel,
FocusList QCheckBox,
FocusList QLabel
```

There is nothing to do with qss itself, it is part of the code, not a bug. This decision, to mix palette roles painting and stylesheets - was mistake. When the Dark theme was made default it simply inherited all these inconsistencies.

VodBox commented:
![Hover Demonstration](https://i.imgur.com/N08CrOM.gif)

The newest commit adds outlines to scene items when hovered over in the sources list, or in the preview. When in the preview, it will highlight the one that will be selected by the next click. This change also results in a thicker outline for selection in general.

SuslikV commented:
And if my source is blue?

VodBox commented:
OBS doesn't exactly have the clearest outlines for sources anyway. The only reason red (or green in cases of crop) works at all is because of the handles. At least this change makes the outlines themselves thicker, uses a color that's distinguishable from red with almost all forms of color blindness, and will still be visible in a lot of cases since it's not pure blue.

Heufneutje commented:
Would it be an idea to show the handles when hovering in that case?

jp9000 commented on libobs/obs.h:
```diff
@@ -594,6 +594,7 @@ enum obs_base_effect {
 	OBS_EFFECT_DEFAULT_RECT,       /**< RGB/YUV (using texture_rect) */
 	OBS_EFFECT_OPAQUE,             /**< RGB/YUV (alpha set to 1.0) */
 	OBS_EFFECT_SOLID,              /**< RGB/YUV (solid color only) */
+	OBS_EFFECT_REPEAT,             /**< RGB/YUV (repeating) */
```
Needs to be at the bottom of the enum, otherwise breaks API
